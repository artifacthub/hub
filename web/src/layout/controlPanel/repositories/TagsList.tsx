import { isEmpty } from 'lodash';
import { nanoid } from 'nanoid';
import { ChangeEvent, Dispatch, MouseEvent as ReactMouseEvent, SetStateAction, useEffect } from 'react';
import { HiPlus } from 'react-icons/hi';
import { MdClose } from 'react-icons/md';

import { ContainerTag } from '../../../types';
import InputField from '../../common/InputField';
import styles from './TagsList.module.css';

interface Props {
  tags: ContainerTag[];
  setContainerTags: Dispatch<SetStateAction<ContainerTag[]>>;
  repeatedTagNames: boolean;
  setRepeatedTagNames: Dispatch<SetStateAction<boolean>>;
}

const EMPTY_TAG = { name: '', mutable: false };

const TagsList = (props: Props) => {
  const cleanRepeatedError = () => {
    if (props.repeatedTagNames) {
      props.setRepeatedTagNames(false);
    }
  };

  const deleteTag = (index: number) => {
    cleanRepeatedError();
    let updatedTags = [...props.tags];
    updatedTags.splice(index, 1);
    props.setContainerTags(updatedTags);
  };

  const addTag = () => {
    props.setContainerTags([...props.tags, { ...EMPTY_TAG, id: nanoid() }]);
  };

  const onUpdateTag = (index: number, field: 'name' | 'mutable', value?: string) => {
    cleanRepeatedError();
    let tagToUpdate: ContainerTag = props.tags[index];
    if (field === 'name') {
      tagToUpdate[field] = value as string;
    } else {
      tagToUpdate[field] = !tagToUpdate.mutable;
    }
    let updatedTags = [...props.tags];
    updatedTags[index] = tagToUpdate;
    props.setContainerTags(updatedTags);
  };

  useEffect(() => {
    if (isEmpty(props.tags)) {
      props.setContainerTags([{ ...EMPTY_TAG, id: nanoid() }]);
    }
  }, [props.tags]); /* eslint-disable-line react-hooks/exhaustive-deps */

  return (
    <div className="mb-4">
      <label className={`form-check-label fw-bold mb-2 ${styles.label}`}>
        Tags
        <button
          type="button"
          className={`btn btn-primary btn-sm ms-2 p-0 position-relative lh-1 ${styles.btn} ${styles.inTitle}`}
          onClick={addTag}
          disabled={props.tags.length === 10}
          aria-label="Add tag"
        >
          <HiPlus />
        </button>
      </label>

      {props.tags.length > 0 && (
        <>
          {props.tags.map((item: ContainerTag, idx: number) => {
            return (
              <div
                className="d-flex flex-row align-items-stretch justify-content-between"
                key={`tag_${item.name}_${idx}`}
              >
                <InputField
                  className="flex-grow-1"
                  type="text"
                  name={`tag_${idx}`}
                  autoComplete="off"
                  value={item.name}
                  placeholder="Tag name"
                  onBlur={(e: ChangeEvent<HTMLInputElement>) => {
                    onUpdateTag(idx, 'name', e.target.value);
                  }}
                  smallBottomMargin
                />

                <div className="d-flex flex-row align-items-center mb-3 ms-3 flex-nowrap">
                  <div className={`ms-2 me-5 position-relative ${styles.inputWrapper}`}>
                    <div className="form-check form-switch ps-0">
                      <label htmlFor={`mutable_${idx}`} className={`form-check-label fw-bold ${styles.label}`}>
                        Mutable
                      </label>
                      <input
                        id={`mutable_${idx}`}
                        type="checkbox"
                        className="form-check-input position-absolute ms-2"
                        role="switch"
                        value="true"
                        checked={item.mutable}
                        onChange={() => {
                          onUpdateTag(idx, 'mutable');
                        }}
                      />
                    </div>
                  </div>

                  <div className={`position-relative text-end ${styles.btnWrapper}`}>
                    <button
                      className={`btn btn-danger btn-sm ms-auto p-0 position-relative lh-1 ${styles.btn}`}
                      type="button"
                      onClick={(event: ReactMouseEvent<HTMLButtonElement, MouseEvent>) => {
                        event.preventDefault();
                        event.stopPropagation();
                        deleteTag(idx);
                      }}
                      aria-label="Delete tag from repository"
                    >
                      <MdClose />
                    </button>
                  </div>
                </div>
              </div>
            );
          })}
        </>
      )}

      {props.repeatedTagNames && <div className="form-text text-danger mt-0">Tags names must be unique.</div>}

      <div className="form-text text-muted mt-2">
        The tags you'd like to list on Artifact Hub must be explicitly added. You can add up to{' '}
        <span className="fw-bold">10</span> (they can be edited later though). Mutable tags will be processed
        periodically, whereas immutable tags will be only processed once.
      </div>
    </div>
  );
};

export default TagsList;
