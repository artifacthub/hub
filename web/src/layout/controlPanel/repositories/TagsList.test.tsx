import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import TagsList from './TagsList';

const setContainerTagsMock = jest.fn();
const setRepeatedTagNamesMock = jest.fn();

const defaultProps = {
  tags: [
    { name: 'tag1', mutable: false },
    { name: 'tag2', mutable: false },
    { name: 'tag3', mutable: true },
  ],
  setContainerTags: setContainerTagsMock,
  repeatedTagNames: false,
  setRepeatedTagNames: setRepeatedTagNamesMock,
};

describe('TagsList', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<TagsList {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      render(<TagsList {...defaultProps} />);

      expect(screen.getByRole('button', { name: 'Add tag' })).toBeInTheDocument();
      expect(screen.getAllByRole('textbox')).toHaveLength(3);
      const checks = screen.getAllByRole('switch');
      expect(checks).toHaveLength(3);
      expect(checks[0]).not.toBeChecked();
      expect(checks[1]).not.toBeChecked();
      expect(checks[2]).toBeChecked();
      expect(
        screen.getByText(/The tags you'd like to list on Artifact Hub must be explicitly added. You can add up to/)
      ).toBeInTheDocument();
    });

    it('updates tag name', async () => {
      render(<TagsList {...defaultProps} />);
      const inputs = screen.getAllByRole('textbox');
      await userEvent.type(inputs[0], 'a');

      act(() => {
        inputs[0].blur();
      });

      await waitFor(() => {
        expect(setContainerTagsMock).toHaveBeenCalledTimes(1);
        expect(setContainerTagsMock).toHaveBeenCalledWith([
          { mutable: false, name: 'tag1a' },
          { mutable: false, name: 'tag2' },
          { mutable: true, name: 'tag3' },
        ]);
      });
    });

    it('updates tag name when repeatedtagnames is true', async () => {
      render(
        <TagsList
          {...defaultProps}
          tags={[
            { name: 'tag1', mutable: false },
            { name: 'tag1', mutable: false },
          ]}
          repeatedTagNames
        />
      );
      const inputs = screen.getAllByRole('textbox');
      await userEvent.type(inputs[0], 'a');

      act(() => {
        inputs[0].blur();
      });

      await waitFor(() => {
        expect(setRepeatedTagNamesMock).toHaveBeenCalledTimes(1);
        expect(setRepeatedTagNamesMock).toHaveBeenCalledWith(false);
      });

      await waitFor(() => {
        expect(setContainerTagsMock).toHaveBeenCalledTimes(1);
        expect(setContainerTagsMock).toHaveBeenCalledWith([
          { mutable: false, name: 'tag1a' },
          { mutable: false, name: 'tag1' },
        ]);
      });
    });

    it('adds tag', async () => {
      render(<TagsList {...defaultProps} />);

      expect(screen.getAllByRole('textbox')).toHaveLength(3);

      const btn = screen.getByRole('button', { name: 'Add tag' });
      userEvent.click(btn);

      await waitFor(() => {
        expect(setContainerTagsMock).toHaveBeenCalledTimes(1);
      });
    });
  });
});
