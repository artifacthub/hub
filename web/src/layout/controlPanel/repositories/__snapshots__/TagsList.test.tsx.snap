// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsList creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="mb-4"
  >
    <label
      class="form-check-label fw-bold mb-2 label"
    >
      Tags
      <button
        aria-label="Add tag"
        class="btn btn-primary btn-sm ms-2 p-0 position-relative lh-1 btn inTitle"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 3a1 1 0 011 1v5h5a1 1 0 110 2h-5v5a1 1 0 11-2 0v-5H4a1 1 0 110-2h5V4a1 1 0 011-1z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </label>
    <div
      class="d-flex flex-row align-items-stretch justify-content-between"
    >
      <div
        class="mb-3 position-relative flex-grow-1"
      >
        <input
          autocomplete="off"
          class="form-control"
          data-testid="tag_0Input"
          id="tag_0"
          name="tag_0"
          placeholder="Tag name"
          spellcheck="false"
          type="text"
          value="tag1"
        />
        <div
          class="invalid-feedback mt-0 inputFeedback"
        />
      </div>
      <div
        class="d-flex flex-row align-items-center mb-3 ms-3 flex-nowrap"
      >
        <div
          class="ms-2 me-5 position-relative inputWrapper"
        >
          <div
            class="form-check form-switch ps-0"
          >
            <label
              class="form-check-label fw-bold label"
              for="mutable_0"
            >
              Mutable
            </label>
            <input
              class="form-check-input position-absolute ms-2"
              id="mutable_0"
              role="switch"
              type="checkbox"
              value="true"
            />
          </div>
        </div>
        <div
          class="position-relative text-end btnWrapper"
        >
          <button
            aria-label="Delete tag from repository"
            class="btn btn-danger btn-sm ms-auto p-0 position-relative lh-1 btn"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="d-flex flex-row align-items-stretch justify-content-between"
    >
      <div
        class="mb-3 position-relative flex-grow-1"
      >
        <input
          autocomplete="off"
          class="form-control"
          data-testid="tag_1Input"
          id="tag_1"
          name="tag_1"
          placeholder="Tag name"
          spellcheck="false"
          type="text"
          value="tag2"
        />
        <div
          class="invalid-feedback mt-0 inputFeedback"
        />
      </div>
      <div
        class="d-flex flex-row align-items-center mb-3 ms-3 flex-nowrap"
      >
        <div
          class="ms-2 me-5 position-relative inputWrapper"
        >
          <div
            class="form-check form-switch ps-0"
          >
            <label
              class="form-check-label fw-bold label"
              for="mutable_1"
            >
              Mutable
            </label>
            <input
              class="form-check-input position-absolute ms-2"
              id="mutable_1"
              role="switch"
              type="checkbox"
              value="true"
            />
          </div>
        </div>
        <div
          class="position-relative text-end btnWrapper"
        >
          <button
            aria-label="Delete tag from repository"
            class="btn btn-danger btn-sm ms-auto p-0 position-relative lh-1 btn"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="d-flex flex-row align-items-stretch justify-content-between"
    >
      <div
        class="mb-3 position-relative flex-grow-1"
      >
        <input
          autocomplete="off"
          class="form-control"
          data-testid="tag_2Input"
          id="tag_2"
          name="tag_2"
          placeholder="Tag name"
          spellcheck="false"
          type="text"
          value="tag3"
        />
        <div
          class="invalid-feedback mt-0 inputFeedback"
        />
      </div>
      <div
        class="d-flex flex-row align-items-center mb-3 ms-3 flex-nowrap"
      >
        <div
          class="ms-2 me-5 position-relative inputWrapper"
        >
          <div
            class="form-check form-switch ps-0"
          >
            <label
              class="form-check-label fw-bold label"
              for="mutable_2"
            >
              Mutable
            </label>
            <input
              checked=""
              class="form-check-input position-absolute ms-2"
              id="mutable_2"
              role="switch"
              type="checkbox"
              value="true"
            />
          </div>
        </div>
        <div
          class="position-relative text-end btnWrapper"
        >
          <button
            aria-label="Delete tag from repository"
            class="btn btn-danger btn-sm ms-auto p-0 position-relative lh-1 btn"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="form-text text-muted mt-2"
    >
      The tags you'd like to list on Artifact Hub must be explicitly added. You can add up to 
      <span
        class="fw-bold"
      >
        10
      </span>
       (they can be edited later though). Mutable tags will be processed periodically, whereas immutable tags will be only processed once.
    </div>
  </div>
</DocumentFragment>
`;
